contrib/rabbitmq/pkg/adapter/adapter.go

```diff
@@ -0,0 +1,262 @@
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package rabbitmq
+
+import (
+	"context"
+	"encoding/json"
+	"github.com/sbcd90/wabbit"
+	"github.com/sbcd90/wabbit/amqp"
+	"github.com/cloudevents/sdk-go/pkg/cloudevents"
+	"github.com/cloudevents/sdk-go/pkg/cloudevents/client"
+	"github.com/cloudevents/sdk-go/pkg/cloudevents/types"
+	"github.com/knative/eventing-sources/pkg/kncloudevents"
+	"github.com/knative/pkg/logging"
+	"go.uber.org/zap"
+	"strings"
+)
+
+const (
+	eventType = "dev.knative.rabbitmq.event"
+)
+
+type ExchangeConfig struct {
+	Name        string
+	TypeOf      string
+	Durable     bool
+	AutoDeleted bool
+	Internal    bool
+	NoWait      bool
+}
+
+type ChannelConfig struct {
+	PrefetchCount int
+	GlobalQos     bool
+}
+
+type QueueConfig struct {
+	Name             string
+	RoutingKey       string
+	Durable          bool
+	DeleteWhenUnused bool
+	Exclusive        bool
+	NoWait           bool
+}
+
+type Adapter struct {
+	Brokers        string
+	Topic          string
+	ChannelConfig  ChannelConfig
+	ExchangeConfig ExchangeConfig
+	QueueConfig    QueueConfig
+	SinkURI        string
+	client         client.Client
+}
+
+func (a *Adapter) initClient() error {
+	if a.client == nil {
+		var err error
+		if a.client, err = kncloudevents.NewDefaultClient(a.SinkURI); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func (a *Adapter) Start(ctx context.Context, stopCh <-chan struct{}) error {
+	logger := logging.FromContext(ctx)
+
+	if err := a.initClient(); err != nil {
+		logger.Error("Failed to create cloudevent client", zap.Error(err))
+		return err
+	}
+
+	logger.Info("Starting with config: ", zap.Any("adapter", a))
+
+	conn, err := amqp.Dial(a.Brokers)
+	if err != nil {
+		logger.Error(err)
+	}
+	defer conn.Close()
+
+	ch, err := conn.Channel()
+	if err != nil {
+		logger.Error(err)
+	}
+	defer ch.Close()
+
+	err = ch.Qos(a.ChannelConfig.PrefetchCount,
+		0,
+		a.ChannelConfig.GlobalQos)
+	if err != nil {
+		logger.Error(err)
+	}
+
+	queue, err := a.StartAmqpClient(ctx, &ch)
+	if err != nil {
+		logger.Error(err)
+	}
+
+	return a.pollForMessages(ctx, &ch, queue, stopCh)
+}
+
+func (a *Adapter) StartAmqpClient(ctx context.Context, ch *wabbit.Channel) (*wabbit.Queue, error) {
+	logger := logging.FromContext(ctx)
+	exchangeConfig := fillDefaultValuesForExchangeConfig(&a.ExchangeConfig, a.Topic)
+
+	err := (*ch).ExchangeDeclare(
+		exchangeConfig.Name,
+		exchangeConfig.TypeOf,
+		wabbit.Option{
+			"durable":  exchangeConfig.Durable,
+			"delete":   exchangeConfig.AutoDeleted,
+			"internal": exchangeConfig.Internal,
+			"noWait":   exchangeConfig.NoWait,
+		},)
```

Comment:
I don't think we need to redeclare the exchange because we are the consumer in that case so the exchange should be created by the producer. We should split the responsibility as explained here: http://gigi.nullneuron.net/gigilabs/rabbitmq-who-creates-the-queues-and-exchanges/. What do you think?